src/datasets/table.py

```diff
@@ -1125,6 +1127,52 @@ def table_cast(table: pa.Table, schema: pa.Schema):
         return table
 
 
+def table_flatten(table: pa.Table):
+    """Improved version of pa.Table.flatten.
+
+    It behaves as pa.Table.flatten, but skips decodable features
+    unless the `decode` attribute of these features is set to False.
+
+    Args:
+        table (Table): PyArrow table to flatten
+
+    Returns:
+        Table: the flattened table
+    """
+    from .features import Audio, Features
+
+    has_decodable_feature = False
+    has_audio_feature = False
+    features = Features.from_arrow_schema(table.schema)
+    for column_name, subfeature in features.items():
+        if hasattr(subfeature, "decode_example") and subfeature.decode:
+            has_decodable_feature = True
+            if isinstance(subfeature, Audio):
+                has_audio_feature = True
```

Comment:
If we have a `flatten` method per type, then we could just check if `subfeature.flatten() == subfeature`. This way we don't have hardcoded lists of feature tyeps that have special handling